### ISSUE TITLE
switch from `bincode` to `rkyv`-based serialization

### ISSUE THREAD

**Issue by blp** [feldera/dbsp#420]
The persistence feature currently uses [`bincode`](https://docs.rs/bincode/latest/bincode/) for serialization and deserialization, as does the [distribution PR](https://github.com/feldera/dbsp/pull/259). This works, but it is difficult to apply to types that come from crates that we don't control, because of the orphan rust for Rust trait implementations. (`bincode` requires two trait implementations: `Encode` and `Decode`.)

This problem already arises for two types: [ArcStr](https://docs.rs/arcstr/latest/arcstr/) and [Decimal](https://docs.rs/rust_decimal/latest/rust_decimal/). For the former, [the maintainer already rejected a `bincode` patch](https://github.com/thomcc/arcstr/pull/45) although he might accept it after `bincode 2.0` stabilizes. For the latter, the situation is [similar](https://github.com/paupino/rust-decimal/pull/589). 

We are currently able to work around these problems by using `#[bincode(with_serde)]` annotations on fields of structs that don't support `bincode`, if they already support `serde`. Both `ArcStr` and `Decimal` are in this category. But this only works for fields in `struct`s that we declare. Lots of other contexts use types in contexts where this doesn't work, which breaks the build. For example, with the current tip of `main`, doing `cargo build --features persistence` yields tons of errors like the following:

```
error[E0277]: the trait bound `ArcStr: Decode` is not satisfied
   --> crates/nexmark/src/queries/q3.rs:43:30
    |
43  |     let person_by_id = input.flat_map_index(|event| match event {
    |                              ^^^^^^^^^^^^^^ the trait `Decode` is not implemented for `ArcStr`
    |
    = help: the following other types implement trait `Decode`:
              ()
              (A, B)
              (A, B, C)
              (A, B, C, D)
              (A, B, C, D, E)
              (A, B, C, D, E, F)
              (A, B, C, D, E, F, G)
              (A, B, C, D, E, F, G, H)
            and 106 others
    = note: required for `(ArcStr, ArcStr, ArcStr)` to implement `Decode`
    = note: required for `(ArcStr, ArcStr, ArcStr)` to implement `DBData`
note: required by a bound in `flat_map_index`
   --> /home/blp/feldera/dbsp/crates/dbsp/src/operator/filter_map.rs:131:12
    |
126 |     fn flat_map_index<F, K, V, I>(&self, func: F) -> Stream<C, OrdIndexedZSet<K, V, Self::R>>
    |        -------------- required by a bound in this associated function
...
131 |         V: DBData,
    |            ^^^^^^ required by this bound in `FilterMap::flat_map_index`

```

For many of these cases, one can create a "newtype" wrapper that implements `Decode` and `Encode`. Unfortunately, this is a big job a lot of the time: take a look at the [300+ lines it took for Decimal](https://github.com/feldera/dbsp/blob/dist5/sql-to-dbsp-compiler/lib/sqllib/src/decimal.rs). And even then, it still doesn't work with the current tree, because sometimes you need to define traits that are impossible to define outside the `dbsp` crate, e.g. [`impl MulByRef for Option<Decimal>`](https://github.com/feldera/dbsp/blob/main/crates/dbsp/src/algebra/mod.rs#L340C1-L347). (It's impossible because `MulByRef` is defined in `dbsp` and `Option` is defined in `std` and the `Decimal` type parameter doesn't factor in.)

Switching from `bincode` to `serde` should ease the situation because `serde` is so ubiquitous that many crates already implement the `serde` traits `Serialize` and `Deserialize`. And the ubiquity of `serde` means that it should be easier to convince crate maintainers to include optional support for it.

**Comment by gz** [comment]
I agree the situation with bincode is not very satisfying (having to implement traits for all types even ones we don't control).

I think there are two main reasons for serialization currently (with maybe a little bit conflicting requirements so that's how we ended up with two serialization frameworks):

- external (dealing with data from clients): here serde seems like a good choice/unavoidable due to maturity and supported features
- internal data (exchange data among hosts, persistence): here we're more flexible and we probably need less features, but performance matters more; I don't have concrete data about what the performance difference (maybe @ryzhyk has more experience). From what I can tell serialize/deserialize at least as of now is a big part of where the time is spent in the current persistence prototype.

- regarding bincode: I picked it initially because we wanted something that's fast
- regarding serde: I think it's slower although it has a lot more features, and we don't have this problem with new types
- another candidate we haven't evaluted (yet): rkyv (supposed to be fast too, but I don't know how it solves this problem?)


Maybe there is a library that is both fast and supports external types better e.g., here is a list https://github.com/djkoloski/rust_serialization_benchmark

**Comment by ryzhyk** [comment]
@gz is correct.  We use `serde` to serialize/deserialize external data.  `serde` is indispensable here, since it supports numerous data formats (csv, json, protobuf, and many others). 

We currently use `bincode` to serialize data for persistent storage and, with #259, to exchange data over the network. In this use case, we don't need to support many formats.  All we care about is performance and size.  The `bincode` crate is supposed to be faster than using `serde` with the bincode format, which I think is why @gz picked it.

In addition, the generality of `serde` comes at the cost of increased compilation time.  At the moment this only affects input and output relations.  If we replace `bincode` with `serde` everywhere, then all intermediate relations will be affected too.

In summary, I am not sure replacing `bincode` with `serde` is a good idea. Replacing `bincode` with `rkyv` or something else _might_ be the right approach, but would have to be based on some performance numbers. Whether we use `bincode`, `rkyv` or something else, it is not going to be as well supported by third-party crates as `serde`, so we will have to build wrappers. I think this is ok long-term, as we only really care about wrappers for types generated by the compiler. This will eventually include some form of interned and/or reference counted types, so we'll need to build wrappers for those, but that's a one-off effort.

**Comment by gz** [comment]
One thing I'm lacking clarity of is: What is this feature that serde has that makes it more convenient to work with non-owned types (is it just that they implemented serialization for all of them in the lib already, or something else)? And if something else does another library (that's also fast) has the same functionality?

**Comment by ryzhyk** [comment]
@gz , why do we care about non-owned types? The way `serde` defines `Deserialize` and `Deserializer` types allows the result of deserialization to point point back to the buffer or string from which it was deserialized, e.g., you can deserialize into `&str` in some cases.  In practice that usually doesn't work because of string escaping, and you need to use `Cow<_, str>` instead. But all our types are owned, so I'm not sure this is something we care about.

**Comment by ryzhyk** [comment]
Or are you referring to remote types from other crates? In that case, yeah, it's just the fact that `serde` is the most popular serialization crate and most other crates that implement data types support it.

**Comment by blp** [comment]
> And even then, it still doesn't work with the current tree, because sometimes you need to define traits that are impossible to define outside the dbsp crate, e.g. [impl MulByRef for Option<Decimal>](https://github.com/feldera/dbsp/blob/main/crates/dbsp/src/algebra/mod.rs#L340C1-L347). (It's impossible because MulByRef is defined in dbsp and Option is defined in std and the Decimal type parameter doesn't factor in.)

Following a suggestion from @Kixiron , I'm working to fix this.

**Comment by ryzhyk** [comment]
@blp , what's the suggestion?

**Comment by blp** [comment]
The suggestion was:

> MulByRef should be implemented for Option<T: MulByRef and then you implement for DecimalWrapper
> Then Option<DecimalWrapper>: MulByRef is satisfied

which boils down to:

```rust
// Implementations of MulByRef for multiplying an `Option` by a value.

trait OptionWeightType {}
impl OptionWeightType for isize {}
impl OptionWeightType for i32 {}
impl OptionWeightType for i64 {}
impl OptionWeightType for f32 {}
impl OptionWeightType for f64 {}
impl OptionWeightType for F32 {}
impl OptionWeightType for F64 {}
impl OptionWeightType for Decimal {}

impl<T, S> MulByRef<S> for Option<T>
where
    T: MulByRef<S, Output = T>,
    S: OptionWeightType,
{
    type Output = Self;

    #[inline]
    fn mul_by_ref(&self, rhs: &S) -> Self::Output {
        self.as_ref().map(|lhs| lhs.mul_by_ref(rhs))
    }
}
```

**Comment by blp** [comment]
I will aim to switch to `rkyv`.

**Comment by blp** [comment]
For DBSP to support a type after `rkyv` transition, it needs to support `SizeOf` and `rkyv`. 

The third-party types I know we need to support are:
- `OrderedFloat`. This type already supports `rkyv` , and `SizeOf` already supports it too.
- `Decimal`. This type already supports `rkyv`, and `SizeOf` already supports it too.
- `ArcStr`. `SizeOf` already support this, but:
  - [ ] `ArcStr` needs `rkyv` support.

**Comment by mihaibudiu** [comment]
I thought the solution was to remove ArcStr

**Comment by blp** [comment]
> I thought the solution was to remove ArcStr

That's an option.

**Comment by gz** [comment]
There was probably a reason (memory consumption, performance, @ryzhyk ?) why the type was chosen over String?

**Comment by mihaibudiu** [comment]
Yes, it's memory consumption.

**Comment by mihaibudiu** [comment]
But the compiler does not generate code with ArcStr. Maybe it should?

**Comment by Kixiron** [comment]
It heavily depends on the workload, I think we should allow users to choose what things get interned

**Comment by mihaibudiu** [comment]
We should be able to make a quick experiment with the compiler.

**Comment by Kixiron** [comment]
> `ArcStr` (or something similar) is probably a no-brainer though, since our strings are immutable.

Agreed, however it's worth noting that some strings within the innards of operators are mutable and we want to have good allocation patterns with those, e.g. formatting strings

**Comment by blp** [comment]
After working on switching to `rkyv` for a while, I think that the correct definition of `DBData` is now something like:

```rust
pub trait DBData<S, D, A>:
    Clone + Eq + Ord + Hash + SizeOf + Send + Debug + Serialize<S> + 'static
where
    S: Fallible + ?Sized,
    D: Fallible + ?Sized,
    A::Archived: Deserialize<Self, D>,
{
}
```

but I sure hope not because adding 3 generic arguments means that lots of other things add 9 generic arguments (for key, value, and weight types) and that's going to be a nightmare.

**Comment by ryzhyk** [comment]
Yes, that'd be terrible.
@Kixiron knows more about `rkyv`. Maybe he can help.

**Comment by blp** [comment]
I might be able to fix some of those types instead of making them generic. I hope so.

**Comment by Kixiron** [comment]
You should just need `rkyv::Archive + rkyv::Deserialize + rkyv::Serialize` (plus maybe a `CheckBytes` bound) on `DBData`. Some things like outputs may need an `rkyv::Archived<T>: serde::Serialize` bound but that shouldn't be needed everywhere

**Comment by Kixiron** [comment]
For the rkyv serializer/deserializer we can also just use a concrete type instead of a generic, we don't actually need that flexibility

**Comment by blp** [comment]
OK, I think that the following will work:

```rust
use rkyv::{ser::serializers::AllocSerializer, Archive, Deserialize, Infallible, Serialize};

pub trait DBData:
    Clone + Eq + Ord + Hash + SizeOf + Send + Debug + Archive + Serialize<Serializer> + 'static
where
    <Self as Archive>::Archived: Deserialize<Self, Deserializer>,
{
}
type Serializer = AllocSerializer<256>;
type Deserializer = Infallible;
```

**Comment by ryzhyk** [comment]
This is a lot better, but I suspect that the `where` clause in this declaration will have to be copied all over the place. I might be wrong.

**Comment by Kixiron** [comment]
We probably want to use a larger buffer than 256 bytes, maybe a `HeapScratch` or something

**Comment by blp** [comment]
> I suspect that the where clause in this declaration will have to be copied all over the place

Yes, and that's a completely monstrous job and the result is very ugly.

I don't understand why Rust doesn't have implied trait bounds.

I suspect that `rkyv` is unusable for this reason.

**Comment by blp** [comment]
An example of the awfulness:

```patch
diff --git a/crates/dbsp/src/operator/time_series/radix_tree/tree_aggregate.rs b/crates/dbsp/src/operator/time_series/radix_tree/tree_aggregate.rs
index f8a289fa..72327f87 100644
--- a/crates/dbsp/src/operator/time_series/radix_tree/tree_aggregate.rs
+++ b/crates/dbsp/src/operator/time_series/radix_tree/tree_aggregate.rs
@@ -170,12 +190,18 @@ where
 impl<Z, IT, OT, Agg, O> TernaryOperator<Z, IT, OT, O> for RadixTreeAggregate<Z, IT, OT, Agg, O>
 where
     Z: IndexedZSet,
+    Archived<Z>: Deserialize<Z, Deserializer>,
     Z::Key: PrimInt,
     IT: BatchReader<Key = Z::Key, Val = Z::Val, Time = (), R = Z::R> + Clone,
+    Archived<IT>: Deserialize<IT, Deserializer>,
     OT: RadixTreeReader<Z::Key, Agg::Accumulator, R = O::R> + Clone,
+    Archived<OT>: Deserialize<OT, Deserializer>,
     Agg: Aggregator<Z::Val, (), Z::R>,
     Agg::Accumulator: Default,
+    Archived<Agg::Accumulator>: Deserialize<Agg::Accumulator, Deserializer>,
+    Archived<Agg::Output>: Deserialize<Agg::Output, Deserializer>,
     O: RadixTreeBatch<Z::Key, Agg::Accumulator>,
+    Archived<O>: Deserialize<O, Deserializer>,
     O::R: ZRingValue,
 {
     fn eval<'a>(
```
